When mask partitions with slice it loses its shape cache

### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any OS, Ray engine
- **Modin version** (`modin.__version__`): a3ddf2f
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})

src_partition = df._query_compiler._modin_frame._partitions[0][0]
masked_partition = src_partition.mask(row_indices=slice(None), col_indices=[0])

assert (
    src_partition._length_cache == masked_partition._length_cache
), f"src: {src_partition._length_cache} | masked: {masked_partition._length_cache}"

```
Output:
```
Traceback (most recent call last):
  File "tst3.py", line 10, in <module>
    ), f"src: {src_partition._length_cache} | masked: {masked_partition._length_cache}"
AssertionError: src: 3 | masked: None
```
<!--
You can obtain the Modin version with

python -c "import modin; print(modin.__version__)"
-->

### Describe the problem
When masking partition with sliced indices the result partition doesn't preserve the shape cache, however, it still can be computed based on the original partition shape. 

Currently, the new shape is cached only when passed indices have `len` property:
https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/engines/ray/pandas_on_ray/frame/partition.py#L150-L153
For slices, if the original partition already has cached shape, we can recompute new shape with the function: [`modin.pandas.indexing.compute_sliced_len`](https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/pandas/indexing.py#L63)

(Issue related with #3100)

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -176,15 +176,11 @@ class PandasFramePartition(ABC):  # pragma: no cover
 
         def try_recompute_cache(indices, previous_cache):
             """Compute new axis-length cache for the masked frame based on its previous cache."""
-            return (
-                (
-                    compute_sliced_len(indices, previous_cache)
-                    if isinstance(previous_cache, int)
-                    else None
-                )
-                if isinstance(indices, slice)
-                else len(indices)
-            )
+            if not isinstance(indices, slice):
+                return len(indices)
+            if not isinstance(previous_cache, int):
+                return None
+            return compute_sliced_len(indices, previous_cache)
 
         new_obj._length_cache = try_recompute_cache(row_indices, self._length_cache)
         new_obj._width_cache = try_recompute_cache(col_indices, self._width_cache)

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -17,7 +17,7 @@ from abc import ABC
 from modin.pandas.indexing import compute_sliced_len
 from copy import copy
 
-import pandas
+from pandas.api.types import is_scalar
 
 
 class PandasFramePartition(ABC):  # pragma: no cover
@@ -140,9 +140,9 @@ class PandasFramePartition(ABC):  # pragma: no cover
 
         Parameters
         ----------
-        row_indices : list-like
+        row_indices : list-like, slice or label
             The indices for the rows to extract.
-        col_indices : list-like
+        col_indices : list-like, slice or label
             The indices for the columns to extract.
 
         Returns
@@ -164,14 +164,15 @@ class PandasFramePartition(ABC):  # pragma: no cover
                 and len(index) == axis_length
             )
 
+        row_indices = [row_indices] if is_scalar(row_indices) else row_indices
+        col_indices = [col_indices] if is_scalar(col_indices) else col_indices
+
         if is_full_axis_mask(row_indices, self._length_cache) and is_full_axis_mask(
             col_indices, self._width_cache
         ):
             return copy(self)
 
-        new_obj = self.add_to_apply_calls(
-            lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])
-        )
+        new_obj = self.add_to_apply_calls(lambda df: df.iloc[row_indices, col_indices])
         new_obj._length_cache = (
             (
                 compute_sliced_len(row_indices, self._length_cache)

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -133,6 +133,17 @@ class PandasFramePartition(ABC):  # pragma: no cover
         """
         pass
 
+    def __copy__(self):
+        """
+        Create a copy of this partition.
+
+        Returns
+        -------
+        PandasFramePartition
+            A copy of this partition.
+        """
+        pass
+
     def mask(self, row_indices, col_indices):
         """
         Lazily create a mask that extracts the indices provided.

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -173,24 +173,21 @@ class PandasFramePartition(ABC):  # pragma: no cover
             return copy(self)
 
         new_obj = self.add_to_apply_calls(lambda df: df.iloc[row_indices, col_indices])
-        new_obj._length_cache = (
-            (
-                compute_sliced_len(row_indices, self._length_cache)
-                if isinstance(self._length_cache, int)
-                else None
+
+        def try_recompute_cache(indices, previous_cache):
+            """Compute new axis-length cache for the masked frame based on its previous cache."""
+            return (
+                (
+                    compute_sliced_len(indices, previous_cache)
+                    if isinstance(previous_cache, int)
+                    else None
+                )
+                if isinstance(indices, slice)
+                else len(indices)
             )
-            if isinstance(row_indices, slice)
-            else len(row_indices)
-        )
-        new_obj._width_cache = (
-            (
-                compute_sliced_len(col_indices, self._width_cache)
-                if isinstance(self._width_cache, int)
-                else None
-            )
-            if isinstance(col_indices, slice)
-            else len(col_indices)
-        )
+
+        new_obj._length_cache = try_recompute_cache(row_indices, self._length_cache)
+        new_obj._width_cache = try_recompute_cache(col_indices, self._width_cache)
         return new_obj
 
     @classmethod

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -14,6 +14,9 @@
 """The module defines base interface for a partition of a Modin DataFrame."""
 
 from abc import ABC
+from modin.pandas.indexing import compute_sliced_len
+
+import pandas
 
 
 class PandasFramePartition(ABC):  # pragma: no cover
@@ -146,7 +149,45 @@ class PandasFramePartition(ABC):  # pragma: no cover
         PandasFramePartition
             New `PandasFramePartition` object.
         """
-        pass
+        if (
+            (isinstance(row_indices, slice) and row_indices == slice(None))
+            or (
+                not isinstance(row_indices, slice)
+                and isinstance(self._length_cache, int)
+                and len(row_indices) == self._length_cache
+            )
+        ) and (
+            (isinstance(col_indices, slice) and col_indices == slice(None))
+            or (
+                not isinstance(col_indices, slice)
+                and isinstance(self._width_cache, int)
+                and len(col_indices) == self._width_cache
+            )
+        ):
+            return self.__copy__()
+
+        new_obj = self.add_to_apply_calls(
+            lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])
+        )
+        new_obj._length_cache = (
+            (
+                compute_sliced_len(row_indices, self._length_cache)
+                if isinstance(self._length_cache, int)
+                else None
+            )
+            if isinstance(row_indices, slice)
+            else len(row_indices)
+        )
+        new_obj._width_cache = (
+            (
+                compute_sliced_len(col_indices, self._width_cache)
+                if isinstance(self._width_cache, int)
+                else None
+            )
+            if isinstance(col_indices, slice)
+            else len(col_indices)
+        )
+        return new_obj
 
     @classmethod
     def put(cls, obj):

--- a/modin/engines/base/frame/partition.py
+++ b/modin/engines/base/frame/partition.py
@@ -15,6 +15,7 @@
 
 from abc import ABC
 from modin.pandas.indexing import compute_sliced_len
+from copy import copy
 
 import pandas
 
@@ -133,17 +134,6 @@ class PandasFramePartition(ABC):  # pragma: no cover
         """
         pass
 
-    def __copy__(self):
-        """
-        Create a copy of this partition.
-
-        Returns
-        -------
-        PandasFramePartition
-            A copy of this partition.
-        """
-        pass
-
     def mask(self, row_indices, col_indices):
         """
         Lazily create a mask that extracts the indices provided.
@@ -160,22 +150,24 @@ class PandasFramePartition(ABC):  # pragma: no cover
         PandasFramePartition
             New `PandasFramePartition` object.
         """
-        if (
-            (isinstance(row_indices, slice) and row_indices == slice(None))
-            or (
-                not isinstance(row_indices, slice)
-                and isinstance(self._length_cache, int)
-                and len(row_indices) == self._length_cache
+
+        def is_full_axis_mask(index, axis_length):
+            """Check whether `index` mask grabs `axis_length` amount of elements."""
+            if isinstance(index, slice):
+                return index == slice(None) or (
+                    isinstance(axis_length, int)
+                    and compute_sliced_len(index, axis_length) == axis_length
+                )
+            return (
+                hasattr(index, "__len__")
+                and isinstance(axis_length, int)
+                and len(index) == axis_length
             )
-        ) and (
-            (isinstance(col_indices, slice) and col_indices == slice(None))
-            or (
-                not isinstance(col_indices, slice)
-                and isinstance(self._width_cache, int)
-                and len(col_indices) == self._width_cache
-            )
+
+        if is_full_axis_mask(row_indices, self._length_cache) and is_full_axis_mask(
+            col_indices, self._width_cache
         ):
-            return self.__copy__()
+            return copy(self)
 
         new_obj = self.add_to_apply_calls(
             lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])

--- a/modin/engines/dask/pandas_on_dask/frame/partition.py
+++ b/modin/engines/dask/pandas_on_dask/frame/partition.py
@@ -24,6 +24,9 @@ from distributed.utils import get_ip
 import cloudpickle as pkl
 from dask.distributed import wait
 
+from distributed.client import _get_global_client
+from modin.pandas.indexing import compute_sliced_len
+
 
 def apply_list_of_funcs(funcs, df):
     """
@@ -174,13 +177,16 @@ class PandasOnDaskFramePartition(PandasFramePartition):
         PandasOnDaskFramePartition
             A new ``PandasOnDaskFramePartition`` object.
         """
-        new_obj = self.add_to_apply_calls(
-            lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])
-        )
-        if not isinstance(row_indices, slice):
-            new_obj._length_cache = len(row_indices)
-        if not isinstance(col_indices, slice):
-            new_obj._width_cache = len(col_indices)
+        new_obj = super().mask(row_indices, col_indices)
+        client = _get_global_client()
+        if isinstance(row_indices, slice) and isinstance(self._length_cache, Future):
+            new_obj._length_cache = client.submit(
+                compute_sliced_len, row_indices, self._length_cache
+            )
+        if isinstance(col_indices, slice) and isinstance(self._width_cache, Future):
+            new_obj._width_cache = client.submit(
+                compute_sliced_len, col_indices, self._width_cache
+            )
         return new_obj
 
     def __copy__(self):

--- a/modin/engines/dask/pandas_on_dask/frame/partition.py
+++ b/modin/engines/dask/pandas_on_dask/frame/partition.py
@@ -167,9 +167,9 @@ class PandasOnDaskFramePartition(PandasFramePartition):
 
         Parameters
         ----------
-        row_indices : list-like
+        row_indices : list-like, slice or label
             The indices for the rows to extract.
-        col_indices : list-like
+        col_indices : list-like, slice or label
             The indices for the columns to extract.
 
         Returns

--- a/modin/engines/python/pandas_on_python/frame/partition.py
+++ b/modin/engines/python/pandas_on_python/frame/partition.py
@@ -48,22 +48,6 @@ class PandasOnPythonFramePartition(PandasFramePartition):
         self.drain_call_queue()
         return self.data.copy()
 
-    def __copy__(self):
-        """
-        Create a copy of this partition.
-
-        Returns
-        -------
-        PandasOnPythonFramePartition
-            A copy of this partition.
-        """
-        return type(self)(
-            data=self.data,
-            length=self._length_cache,
-            width=self._width_cache,
-            call_queue=self.call_queue,
-        )
-
     def apply(self, func, **kwargs):
         """Apply some callable function to the data in this partition.
 

--- a/modin/engines/python/pandas_on_python/frame/partition.py
+++ b/modin/engines/python/pandas_on_python/frame/partition.py
@@ -48,6 +48,22 @@ class PandasOnPythonFramePartition(PandasFramePartition):
         self.drain_call_queue()
         return self.data.copy()
 
+    def __copy__(self):
+        """
+        Create a copy of this partition.
+
+        Returns
+        -------
+        PandasOnPythonFramePartition
+            A copy of this partition.
+        """
+        return type(self)(
+            data=self.data,
+            length=self._length_cache,
+            width=self._width_cache,
+            call_queue=self.call_queue,
+        )
+
     def apply(self, func, **kwargs):
         """Apply some callable function to the data in this partition.
 

--- a/modin/engines/python/pandas_on_python/frame/partition.py
+++ b/modin/engines/python/pandas_on_python/frame/partition.py
@@ -90,16 +90,6 @@ class PandasOnPythonFramePartition(PandasFramePartition):
     def wait(self):
         self.drain_call_queue()
 
-    def mask(self, row_indices=None, col_indices=None):
-        new_obj = self.add_to_apply_calls(
-            lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])
-        )
-        if not isinstance(row_indices, slice):
-            new_obj._length_cache = len(row_indices)
-        if not isinstance(col_indices, slice):
-            new_obj._width_cache = len(col_indices)
-        return new_obj
-
     def to_pandas(self):
         """Convert the object stored in this partition to a Pandas DataFrame.
 

--- a/modin/engines/ray/pandas_on_ray/frame/partition.py
+++ b/modin/engines/ray/pandas_on_ray/frame/partition.py
@@ -18,6 +18,7 @@ import pandas
 from modin.data_management.utils import length_fn_pandas, width_fn_pandas
 from modin.engines.base.frame.partition import PandasFramePartition
 from modin.engines.ray.utils import handle_ray_task_error
+from modin.pandas.indexing import compute_sliced_len
 
 import ray
 from ray.worker import RayTaskError
@@ -29,6 +30,8 @@ if version.parse(ray.__version__) >= version.parse("1.2.0"):
     from ray.util.client.common import ClientObjectRef
 
     ObjectIDType = (ray.ObjectRef, ClientObjectRef)
+
+compute_sliced_len = ray.remote(compute_sliced_len)
 
 
 class PandasOnRayFramePartition(PandasFramePartition):
@@ -209,30 +212,19 @@ class PandasOnRayFramePartition(PandasFramePartition):
         PandasOnRayFramePartition
             A new ``PandasOnRayFramePartition`` object.
         """
-        if (
-            (isinstance(row_indices, slice) and row_indices == slice(None))
-            or (
-                not isinstance(row_indices, slice)
-                and self._length_cache is not None
-                and len(row_indices) == self._length_cache
-            )
-        ) and (
-            (isinstance(col_indices, slice) and col_indices == slice(None))
-            or (
-                not isinstance(col_indices, slice)
-                and self._width_cache is not None
-                and len(col_indices) == self._width_cache
-            )
+        new_obj = super().mask(row_indices, col_indices)
+        if isinstance(row_indices, slice) and isinstance(
+            self._length_cache, ObjectIDType
         ):
-            return self.__copy__()
-
-        new_obj = self.add_to_apply_calls(
-            lambda df: pandas.DataFrame(df.iloc[row_indices, col_indices])
-        )
-        if not isinstance(row_indices, slice):
-            new_obj._length_cache = len(row_indices)
-        if not isinstance(col_indices, slice):
-            new_obj._width_cache = len(col_indices)
+            new_obj._length_cache = compute_sliced_len.remote(
+                row_indices, self._length_cache
+            )
+        if isinstance(col_indices, slice) and isinstance(
+            self._width_cache, ObjectIDType
+        ):
+            new_obj._width_cache = compute_sliced_len.remote(
+                col_indices, self._width_cache
+            )
         return new_obj
 
     @classmethod

--- a/modin/engines/ray/pandas_on_ray/frame/partition.py
+++ b/modin/engines/ray/pandas_on_ray/frame/partition.py
@@ -202,9 +202,9 @@ class PandasOnRayFramePartition(PandasFramePartition):
 
         Parameters
         ----------
-        row_indices : list-like
+        row_indices : list-like, slice or label
             The indices for the rows to extract.
-        col_indices : list-like
+        col_indices : list-like, slice or label
             The indices for the columns to extract.
 
         Returns

--- a/modin/pandas/test/test_api.py
+++ b/modin/pandas/test/test_api.py
@@ -45,6 +45,7 @@ def test_top_level_api_equality():
     ]
 
     ignore_modin = [
+        "indexing",
         "iterator",
         "series",
         "accessor",

--- a/modin/test/test_partition_api.py
+++ b/modin/test/test_partition_api.py
@@ -19,14 +19,35 @@ import modin.pandas as pd
 from modin.distributed.dataframe.pandas import unwrap_partitions, from_partitions
 from modin.config import Engine, NPartitions
 from modin.pandas.test.utils import df_equals
+from modin.pandas.indexing import compute_sliced_len
 
+from modin.data_management.factories.dispatcher import EngineDispatcher
+
+PartitionClass = (
+    EngineDispatcher.get_engine().io_cls.frame_cls._partition_mgr_cls._partition_class
+)
 
 if Engine.get() == "Ray":
     import ray
-if Engine.get() == "Dask":
+
+    put_func = ray.put
+    get_func = ray.get
+    FutureType = ray.ObjectRef
+elif Engine.get() == "Dask":
     from distributed.client import get_client
+    from distributed import Future
+
+    put_func = lambda x: get_client().scatter(x)  # noqa: E731
+    get_func = lambda x: x.result()  # noqa: E731
+    FutureType = Future
+else:
+    put_func = lambda x: x  # noqa: E731
+    get_func = lambda x: x  # noqa: E731
+    FutureType = object
 
 NPartitions.put(4)
+# HACK: implicit engine initialization (Modin issue #2989)
+pd.DataFrame([])
 
 
 @pytest.mark.parametrize("axis", [None, 0, 1])
@@ -96,3 +117,53 @@ def test_from_partitions(axis):
             futures = client.scatter([df1, df2], hash=False)
     actual_df = from_partitions(futures, axis)
     df_equals(expected_df, actual_df)
+
+
+@pytest.mark.parametrize("row_indices", [[0, 2], slice(None, None, 2)])
+@pytest.mark.parametrize("col_indices", [[0, 2], slice(None, None, 2)])
+@pytest.mark.parametrize("is_length_future", [False, True])
+@pytest.mark.parametrize("is_width_future", [False, True])
+def test_mask_preserve_cache(
+    row_indices, col_indices, is_length_future, is_width_future
+):
+    def serialize(obj):
+        if isinstance(obj, FutureType):
+            return get_func(obj)
+        return obj
+
+    def compute_length(indices, length):
+        if not isinstance(indices, slice):
+            return len(indices)
+        return compute_sliced_len(indices, length)
+
+    df = pandas.DataFrame({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8], "c": [9, 10, 11, 12]})
+    obj_id = put_func(df)
+
+    partition_shape = [
+        put_func(len(df)) if is_length_future else len(df),
+        put_func(len(df.columns)) if is_width_future else len(df.columns),
+    ]
+
+    source_partition = PartitionClass(obj_id, *partition_shape)
+    masked_partition = source_partition.mask(
+        row_indices=row_indices, col_indices=col_indices
+    )
+
+    expected_length = compute_length(row_indices, len(df))
+    expected_width = compute_length(col_indices, len(df.columns))
+
+    # Check that the cache is preserved
+    assert expected_length == serialize(masked_partition._length_cache)
+    assert expected_width == serialize(masked_partition._width_cache)
+    # Check that the cache is interpreted properly
+    assert expected_length == masked_partition.length()
+    assert expected_width == masked_partition.width()
+    # Recompute shape explicitly to check that the cached data was correct
+    expected_length, expected_width = [
+        masked_partition._length_cache,
+        masked_partition._width_cache,
+    ]
+    masked_partition._length_cache = None
+    masked_partition._width_cache = None
+    assert expected_length == masked_partition.length()
+    assert expected_width == masked_partition.width()

--- a/modin/test/test_partition_api.py
+++ b/modin/test/test_partition_api.py
@@ -20,7 +20,6 @@ from modin.distributed.dataframe.pandas import unwrap_partitions, from_partition
 from modin.config import Engine, NPartitions
 from modin.pandas.test.utils import df_equals
 from modin.pandas.indexing import compute_sliced_len
-
 from modin.data_management.factories.dispatcher import EngineDispatcher
 
 PartitionClass = (
@@ -119,8 +118,8 @@ def test_from_partitions(axis):
     df_equals(expected_df, actual_df)
 
 
-@pytest.mark.parametrize("row_indices", [[0, 2], slice(None, None, 2)])
-@pytest.mark.parametrize("col_indices", [[0, 2], slice(None, None, 2)])
+@pytest.mark.parametrize("row_indices", [[0, 2], slice(None)])
+@pytest.mark.parametrize("col_indices", [[0, 2], slice(None)])
 @pytest.mark.parametrize("is_length_future", [False, True])
 @pytest.mark.parametrize("is_width_future", [False, True])
 def test_mask_preserve_cache(

--- a/modin/test/test_partition_api.py
+++ b/modin/test/test_partition_api.py
@@ -39,10 +39,14 @@ elif Engine.get() == "Dask":
     put_func = lambda x: get_client().scatter(x)  # noqa: E731
     get_func = lambda x: x.result()  # noqa: E731
     FutureType = Future
-else:
+elif Engine.get() == "Python":
     put_func = lambda x: x  # noqa: E731
     get_func = lambda x: x  # noqa: E731
     FutureType = object
+else:
+    raise NotImplementedError(
+        f"'{Engine.get()}' engine is not supported by these test suites"
+    )
 
 NPartitions.put(4)
 # HACK: implicit engine initialization (Modin issue #2989)

--- a/modin/test/test_partition_api.py
+++ b/modin/test/test_partition_api.py
@@ -125,7 +125,7 @@ def test_from_partitions(axis):
 def test_mask_preserve_cache(
     row_indices, col_indices, is_length_future, is_width_future
 ):
-    def serialize(obj):
+    def deserialize(obj):
         if isinstance(obj, FutureType):
             return get_func(obj)
         return obj
@@ -152,8 +152,8 @@ def test_mask_preserve_cache(
     expected_width = compute_length(col_indices, len(df.columns))
 
     # Check that the cache is preserved
-    assert expected_length == serialize(masked_partition._length_cache)
-    assert expected_width == serialize(masked_partition._width_cache)
+    assert expected_length == deserialize(masked_partition._length_cache)
+    assert expected_width == deserialize(masked_partition._width_cache)
     # Check that the cache is interpreted properly
     assert expected_length == masked_partition.length()
     assert expected_width == masked_partition.width()

